File: engine/runtime/src/main/scala/org/enso/compiler/Cache.scala
package org.enso.compiler

import com.oracle.truffle.api.{TruffleFile, TruffleLogger}
import org.bouncycastle.jcajce.provider.digest.SHA3
import org.bouncycastle.util.encoders.Hex
import org.enso.interpreter.runtime.EnsoContext
import org.enso.logger.masking.MaskedPath

import java.io.{
  ByteArrayInputStream,
  ByteArrayOutputStream,
  IOException,
  ObjectInputStream,
  ObjectOutputStream
}
import java.nio.file.{
  FileAlreadyExistsException,
  NoSuchFileException,
  Path,
  StandardOpenOption
}
import java.security.MessageDigest
import java.util.logging.Level
import scala.util.{Failure, Success, Try, Using}

/** Responsible for the management of caches associated with a given data.
  * @tparam T the type of the underlying cached data entry
  */
trait Cache[T] {

  def logLevel: Level

  import Cache._

  /** Type of Metadata representation corresponding to the given data
    */
  type M <: Metadata

  /** Saves the provided IR into the cache for this module.
    *
    * @param entry the data representation to be saved
    * @param context the language context in which saving is taking place
    * @param useGlobalCacheLocations if true, will use global caches' location, local one otherwise
    * @return returns the location of the cache if successful, and [[None]] if
    *         it was unable to save
    */
  def save(
    entry: T,
    context: EnsoContext,
    useGlobalCacheLocations: Boolean
  ): Option[TruffleFile] = {
    implicit val logger: TruffleLogger = context.getLogger(this.getClass)
    getCacheRoots(context) match {
      case Some(roots) =>
        if (useGlobalCacheLocations) {
          if (saveCacheTo(roots.globalCacheRoot, entry)) {
            return Some(roots.globalCacheRoot)
          }
        } else {
          logger.log(
            logLevel,
            s"Skipping use of global cache locations for $stringRepr."
          )
        }

        if (saveCacheTo(roots.localCacheRoot, entry)) {
          return Some(roots.localCacheRoot)
        }

        logger.log(
          logLevel,
          s"Unable to write cache data for $stringRepr."
        )
        None
      case None => None
    }
  }

  protected def computeDigest(entry: T): Option[String]

  protected def computeDigestFromSource(context: EnsoContext): Option[String]

  protected def getCacheRoots(context: EnsoContext): Option[Roots]

  protected def extractObjectToSerialize(entry: T): Object

  /** Saves the cache into the provided `cacheRoot`.
    *
    * @param cacheRoot the root into which the cache should be saved
    * @param entry    the data that should be cached
    * @param logger   a logger
    * @return `true` if the cache was written successfully, `false` otherwise
    */
  private def saveCacheTo(
    cacheRoot: TruffleFile,
    entry: T
  )(implicit logger: TruffleLogger): Boolean = {
    if (ensureRoot(cacheRoot)) {
      val byteStream: ByteArrayOutputStream = new ByteArrayOutputStream()
      val bytesToWrite =
        Using.resource(new ObjectOutputStream(byteStream)) { stream =>
          stream.writeObject(extractObjectToSerialize(entry))
          byteStream.toByteArray
        }

      val blobDigest = computeDigestFromBytes(bytesToWrite)
      val sourceDigest = computeDigest(entry).getOrElse(
        throw new IllegalStateException("unable to compute digest")
      )
      val metadataBytes = metadata(sourceDigest, blobDigest, entry)

      val cacheDataFile = getCacheDataPath(cacheRoot)
      val metadataFile  = getCacheMetadataPath(cacheRoot)
      val parentPath    = cacheDataFile.getParent

      if (writeBytesTo(cacheDataFile, bytesToWrite)) {
        if (writeBytesTo(metadataFile, metadataBytes)) {
          logger.log(
            logLevel,
            s"Written cache data [$stringRepr] to [${parentPath.toMaskedPath.applyMasking()}]."
          )
          return true
        } else {
          // Clean up after ourselves if it fails.
          cacheDataFile.delete()
        }
      }

      return true
    }

    false
  }

  /** Loads the cache for this data where possible.
    *
    * @param context the language context in which loading is taking place
    * @return the cached data if possible, and [[None]] if it could not
    *         load a valid cache
    */
  def load(context: EnsoContext): Option[T] = {
    this.synchronized {
      implicit val logger: TruffleLogger = context.getLogger(this.getClass)
      logger.log(logLevel, "loading from Cache")
      getCacheRoots(context) match {
        case Some(roots) =>
          logger.log(logLevel, "got roots: " + roots)
          // Load from the global root as a priority.
          loadCacheFrom(roots.globalCacheRoot, context) match {
            case cache @ Some(_) =>
              logger.log(
                logLevel,
                s"Using cache for " +
                s"[${stringRepr}] at location " +
                s"[${roots.globalCacheRoot.toMaskedPath.applyMasking()}]."
              )
              return cache
            case None =>
          }

          // Load from the local root if the global root fails.
          loadCacheFrom(roots.localCacheRoot, context) match {
            case cache @ Some(_) =>
              logger.log(
                logLevel,
                s"Using cache for " +
                s"[${stringRepr}] at location " +
                s"[${roots.localCacheRoot.toMaskedPath.applyMasking()}]."
              )
              return cache
            case None =>
          }

          logger.log(
            logLevel,
            s"Unable to load a cache for module [${stringRepr}]"
          )

          None
        case None => None
      }
    }
  }

  /** Loads the cache from the provided `cacheRoot`, invalidating the cache if
    * the loading fails for any reason.
    *
    * @param cacheRoot the root at which to find the cache for this cache entry
    * @param logger    a logger
    * @return the cached data if available, otherwise [[None]].
    */
  private def loadCacheFrom(cacheRoot: TruffleFile, context: EnsoContext)(
    implicit logger: TruffleLogger
  ): Option[T] = {
    val metadataPath = getCacheMetadataPath(cacheRoot)
    val dataPath     = getCacheDataPath(cacheRoot)

    loadCacheMetadata(metadataPath) match {
      case Some(meta) =>
        val sourceDigestValid =
          computeDigestFromSource(context)
            .map(_ == meta.sourceHash)
            .getOrElse(false)
        val blobBytes       = dataPath.readAllBytes()
        val blobDigestValid = computeDigestFromBytes(blobBytes) == meta.blobHash

        if (sourceDigestValid && blobDigestValid) {
          val readObject =
            Using(new ObjectInputStream(new ByteArrayInputStream(blobBytes))) {
              _.readObject()
            }

          readObject match {
            case Success(obj) =>
              validateReadObject(obj, meta) match {
                case Failure(_) =>
                  logger.log(
                    logLevel,
                    s"${stringRepr}` was corrupt on disk."
                  )
                  invalidateCache(cacheRoot)
                  None
                case success =>
                  success.toOption
              }
            case Failure(ex) =>
              logger.log(
                logLevel,
                s"`${stringRepr}` failed to load " +
                s"(caused by: ${ex.getMessage})."
              )
              invalidateCache(cacheRoot)
              None
          }
        } else {
          logger.log(
            logLevel,
            s"One or more digests did not match for the cache for " +
            s"[${stringRepr}]."
          )
          invalidateCache(cacheRoot)
          None
        }
      case None =>
        logger.log(
          logLevel,
          s"Could not load the cache metadata " +
          s"at [${metadataPath.toMaskedPath.applyMasking()}]"
        )
        invalidateCache(cacheRoot)
        None
    }
  }

  /** Validates the deserialized data by returning the expected cached entry, or [[Failure]].
    *
    * @param obj deserialized object
    * @param meta metadata corresponding to the `obj`
    * @return an `obj` transformed to a cached entry, or [[Failure]]
    */
  protected def validateReadObject(obj: AnyRef, meta: M): Try[T]

  /** Read metadata representation from the provided location
    * @param path location of the serialized metadata
    * @return deserialized metadata, or [[None]] if invalid
    */
  private def loadCacheMetadata(
    path: TruffleFile
  ): Option[M] = {
    if (path.isReadable) {
      val bytes = path.readAllBytes()
      metadataFromBytes(bytes)
    } else {
      None
    }
  }

  protected def metadataFromBytes(bytes: Array[Byte]): Option[M]

  private def ensureRoot(cacheRoot: TruffleFile): Boolean = {
    try {
      if (cacheRoot.exists() && cacheRoot.isDirectory()) {
        return cacheRoot.isWritable
      } else {
        cacheRoot.createDirectories()
        return cacheRoot.isWritable
      }
    } catch {
      case _: IOException | _: UnsupportedOperationException |
          _: SecurityException | _: FileAlreadyExistsException =>
    }
    false
  }

  /** Computes SHA-224 digest from the provided bytes.
    */
  protected def computeDigestFromBytes(bytes: Array[Byte]): String = {
    Hex.toHexString(messageDigest().digest(bytes))
  }

  /** Raw (byte)  representation of the [[Metadata]].
    *
    * @param sourceDigest SHA-256 digest of source data
    * @param blobDigest SHA-256 digest of serialized data
    * @param entry cache entry
    * @return serialized metadata
    */
  protected def metadata(
    sourceDigest: String,
    blobDigest: String,
    entry: T
  ): Array[Byte]

  /** Gets the path to the cache data within the `cacheRoot`.
    *
    * @param cacheRoot the root of the cache for this entry
    * @return the name of the data file for this entry's cache
    */
  private def getCacheDataPath(cacheRoot: TruffleFile): TruffleFile = {
    cacheRoot.resolve(cacheFileName(dataSuffix))
  }

  private def getCacheMetadataPath(cacheRoot: TruffleFile): TruffleFile =
    cacheRoot.resolve(cacheFileName(metadataSuffix))

  /** Computes the cache file name for a given extension.
    *
    * @param suffix the extension
    * @return the cache file name with the provided `ext`
    */
  private def cacheFileName(suffix: String): String = {
    s"$entryName$suffix"
  }

  protected def stringRepr: String

  protected def entryName: String

  protected def dataSuffix: String

  protected def metadataSuffix: String

  protected def messageDigest(): MessageDigest = {
    new SHA3.Digest224()
  }

  /** Deletes the cache for this data in the provided `cacheRoot`.
    *
    * @param cacheRoot the root of the cache to delete
    * @param logger    a logger
    */
  private def invalidateCache(
    cacheRoot: TruffleFile
  )(implicit logger: TruffleLogger): Unit = {
    val metadataFile = getCacheMetadataPath(cacheRoot)
    val dataFile     = getCacheDataPath(cacheRoot)

    def doDeleteAt(file: TruffleFile): Unit = {
      try {
        if (file.exists()) {
          if (file.isWritable) {
            file.delete()
            logger.log(
              logLevel,
              s"Invalidated the cache at [${file.toMaskedPath.applyMasking()}]."
            )
          } else {
            logger.log(
              logLevel,
              s"Cannot invalidate the cache at " +
              s"[${file.toMaskedPath.applyMasking()}]. " +
              s"Cache location not writable."
            )
          }
        }
      } catch {
        case _: NoSuchFileException =>
        // If it doesn't exist, our work has already been done for us!
        case _: java.nio.file.DirectoryNotEmptyException | _: IOException |
            _: SecurityException =>
          logger.log(
            logLevel,
            s"Unable to delete the cache at " +
            s"[${cacheRoot.toMaskedPath.applyMasking()}]."
          )
      }
    }

    doDeleteAt(metadataFile)
    doDeleteAt(dataFile)
  }

  /** Invalidates all caches associated with this cache.
    *
    * @param context the langage context in which loading is taking place
    */
  def invalidate(context: EnsoContext): Unit = {
    this.synchronized {
      implicit val logger: TruffleLogger = context.getLogger(this.getClass)
      getCacheRoots(context).foreach { roots =>
        invalidateCache(roots.globalCacheRoot)
        invalidateCache(roots.localCacheRoot)
      }
    }
  }

}

object Cache {

  /** A representation of the cache roots. */
  case class Roots(localCacheRoot: TruffleFile, globalCacheRoot: TruffleFile)

  /** Write the provided `bytes` to the provided `file`.
    *
    * @param file  the file to write the bytes to
    * @param bytes the bytes to write into `file`
    * @return `true` if writing completed successfully, `false` otherwise
    */
  private def writeBytesTo(
    file: TruffleFile,
    bytes: Array[Byte]
  ): Boolean = {
    try {
      Using.resource(
        file.newOutputStream(
          StandardOpenOption.WRITE,
          StandardOpenOption.CREATE,
          StandardOpenOption.TRUNCATE_EXISTING
        )
      ) { stream =>
        stream.write(bytes)
      }

      true
    } catch {
      case _: IOException | _: SecurityException => false
    }
  }

  /** Internal storage for the cache metadata to enable easy way to verify de-serialisation.
    *
    * @param sourceHash       the hex-encoded SHA3-224 hash of the source data for
    *                         which the cache was created
    * @param blobHash         the hex-encoded SHA3-224 hash of the serialized blob as it was
    *                         written to disk
    */
  trait Metadata {
    def sourceHash: String
    def blobHash:   String
  }

  /** Convert a [[TruffleFile]] to a [[MaskedPath]] for the purposes of logging.
    *
    * @param truffleFile the file to convert.
    */
  implicit private class ToMaskedPath(truffleFile: TruffleFile) {
    def toMaskedPath: MaskedPath = {
      MaskedPath(Path.of(truffleFile.getPath))
    }
  }

}

File: engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java
package org.enso.interpreter.caches;

import com.oracle.truffle.api.TruffleFile;
import com.oracle.truffle.api.TruffleLogger;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.OutputStream;
import java.lang.foreign.Arena;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.channels.FileChannel.MapMode;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.logging.Level;
import org.enso.interpreter.runtime.EnsoContext;
import org.enso.logger.masking.MaskedPath;

/**
 * Cache encapsulates a common functionality needed to serialize and de-serialize objects, while
 * maintaining its integrity in the process.
 *
 * @param <T> type of the cached data
 * @param <M> type of the metadata associated with the data
 */
public final class Cache<T, M> {
  private final Object LOCK = new Object();

  /** implementation of the serialize/deserialize operations */
  private final Spi<T, M> spi;

  /** Returns a default level of logging for this Cache. */
  private final Level logLevel;

  /** Log name to use in log messages */
  private final String logName;

  /**
   * Flag indicating if the de-serialization process should compute the hash of the sources from
   * which the cache was created and compare it with the stored metadata entry.
   */
  private final boolean needsSourceDigestVerification;

  /**
   * Large cache files will be {@link FileChannel#map(MapMode, long, long, Arena) mmapped} using a
   * newly created arena via this supplier. Whenever a cache is loaded or saved, the previous arena
   * will be closed, which will invalidate all byte buffers associated with that arena.
   *
   * <p>Note that currently, it is not possible to use {@link Arena#ofConfined()} here. See <a
   * href="https://github.com/enso-org/enso/pull/13872#discussion_r2313983664">GH discussion</a>.
   */
  private final Supplier<Arena> memoryArenaSupplier;

  private Arena memoryArena;

  /**
   * Flag indicating if the de-serialization process should compute the hash of the stored cache and
   * compare it with the stored metadata entry.
   */
  private final boolean needsDataDigestVerification;

  /**
   * Constructor for subclasses.
   *
   * @param logLevel logging level
   * @param logName name to use in logs
   * @param needsSourceDigestVerification Flag indicating if the de-serialization process should
   *     compute the hash of the sources from which the cache was created and compare it with the
   *     stored metadata entry.
   * @param needsDataDigestVerification Flag indicating if the de-serialization process should
   *     compute the hash of the stored cache and compare it with the stored metadata entry.
   */
  private Cache(
      Spi<T, M> spi,
      Level logLevel,
      String logName,
      boolean needsSourceDigestVerification,
      boolean needsDataDigestVerification,
      Supplier<Arena> memoryArenaSupplier) {
    this.spi = spi;
    this.logLevel = logLevel;
    this.logName = logName;
    this.needsDataDigestVerification = needsDataDigestVerification;
    this.needsSourceDigestVerification = needsSourceDigestVerification;
    this.memoryArenaSupplier = memoryArenaSupplier;
  }

  /**
   * Factory method to create new cache instance.
   *
   * @param spi the implementation logic of the cache
   * @param logLevel logging level
   * @param logName name to use in logs
   * @param needsSourceDigestVerification Flag indicating if the de-serialization process should
   *     compute the hash of the sources from which the cache was created and compare it with the
   *     stored metadata entry.
   * @param needsDataDigestVerification Flag indicating if the de-serialization process should
   *     compute the hash of the stored cache and compare it with the stored metadata entry.
   */
  static <T, M> Cache<T, M> create(
      Spi<T, M> spi,
      Level logLevel,
      String logName,
      boolean needsSourceDigestVerification,
      boolean needsDataDigestVerification) {
    return new Cache<>(
        spi,
        logLevel,
        logName,
        needsSourceDigestVerification,
        needsDataDigestVerification,
        Arena::ofShared);
  }

  static <T, M> Cache<T, M> create(
      Spi<T, M> spi,
      Level logLevel,
      String logName,
      boolean needsSourceDigestVerification,
      boolean needsDataDigestVerification,
      Supplier<Arena> memoryArenaSupplier) {
    return new Cache<>(
        spi,
        logLevel,
        logName,
        needsSourceDigestVerification,
        needsDataDigestVerification,
        memoryArenaSupplier);
  }

  /**
   * Saves data to a cache file.
   *
   * @param entry data to save
   * @param context the language context in which loading is taking place
   * @return the location of the successfully saved location of the cached data
   * @throws IOException if something goes wrong
   */
  public final TruffleFile save(T entry, EnsoContext context) throws IOException {
    TruffleLogger logger = context.getLogger(this.getClass());
    for (var root : spi.getCacheRoots(context)) {
      if (saveCacheTo(context, root, entry, logger)) {
        return root;
      }
    }
    throw new IOException("Unable to write cache data for " + logName + ".");
  }

  /**
   * Attempts to save cache data at a specified location.
   *
   * @param cacheRoot parent directory where cache data should be stored
   * @param entry cache data to save
   * @param logger internal logger
   * @return true, if successful, false otherwise
   * @throws IOException IOException encountered while writing data to files
   */
  private boolean saveCacheTo(
      EnsoContext context, TruffleFile cacheRoot, T entry, TruffleLogger logger)
      throws IOException {
    if (ensureRoot(cacheRoot)) {
      byte[] bytesToWrite = spi.serialize(context, entry);

      String blobDigest = CacheUtils.computeDigestFromBytes(ByteBuffer.wrap(bytesToWrite));
      String sourceDigest = spi.computeDigest(entry, logger).get();
      if (sourceDigest == null) {
        throw new IOException("unable to compute digest");
      }
      byte[] metadataBytes = spi.metadata(sourceDigest, blobDigest, entry);

      TruffleFile cacheDataFile = getCacheDataPath(cacheRoot);
      TruffleFile metadataFile = getCacheMetadataPath(cacheRoot);
      TruffleFile parentPath = cacheDataFile.getParent();

      closeMemoryArena();
      if (writeBytesTo(cacheDataFile, bytesToWrite) && writeBytesTo(metadataFile, metadataBytes)) {
        logger.log(
            logLevel,
            "Written cache data [{0}] to [{1}] of size [{2}].",
            new Object[] {logName, toMaskedPath(parentPath).applyMasking(), bytesToWrite.length});
        return true;
      } else {
        // Clean up after ourselves if it fails.
        cacheDataFile.delete();
      }
    }
    return false;
  }

  private boolean ensureRoot(TruffleFile cacheRoot) {
    try {
      if (cacheRoot.exists() && cacheRoot.isDirectory()) {
        return cacheRoot.isWritable();
      } else {
        cacheRoot.createDirectories();
        return cacheRoot.isWritable();
      }
    } catch (Throwable e) {
      return false;
    }
  }

  /**
   * Loads cache for this data, if possible.
   *
   * @param context the language context in which loading is taking place
   * @return the cached data if possible, and [[None]] if it could not load a valid cache
   */
  public final Optional<T> load(EnsoContext context) {
    var logger = context.getLogger(this.getClass());
    var collected = new ArrayList<IOException>();
    synchronized (LOCK) {
      for (var root : spi.getCacheRoots(context)) {
        try {
          var cache = loadCacheFrom(root, context, logger);
          logger.log(
              logLevel,
              "Using cache for ["
                  + logName
                  + "] at location ["
                  + toMaskedPath(root).applyMasking()
                  + "].");
          return Optional.of(cache);
        } catch (IOException ex) {
          collected.add(ex);
        }
      }
    }
    for (var ex : collected) {
      logCacheLoadFailure(logger, "Unable to load a cache [" + logName + "]: ", ex);
    }
    return Optional.empty();
  }

  private void logCacheLoadFailure(TruffleLogger logger, String prefix, IOException ex) {
    if (ex instanceof FileNotFoundException) {
      logger.log(Level.FINE, prefix + ex.getMessage());
    } else {
      logger.log(Level.FINE, prefix + ex.getMessage(), ex);
    }
  }

  /**
   * Loads the cache from the provided `cacheRoot`, invalidating the cache if the loading fails for
   * any reason.
   *
   * @param cacheRoot the root at which to find the cache for this cache entry
   * @param context the language context in which loading is taking place
   * @param logger a logger
   * @return the cached data if available, otherwise throw exception
   */
  private T loadCacheFrom(TruffleFile cacheRoot, EnsoContext context, TruffleLogger logger)
      throws IOException {
    TruffleFile metadataPath = getCacheMetadataPath(cacheRoot);
    TruffleFile dataPath = getCacheDataPath(cacheRoot);

    var meta = loadCacheMetadata(metadataPath, logger);
    if (meta != null) {
      boolean sourceDigestValid =
          !needsSourceDigestVerification
              || spi.computeDigestFromSource(context, logger)
                  .map(digest -> digest.equals(spi.sourceHash(meta)))
                  .orElseGet(() -> false);
      var file = new File(dataPath.toUri());
      ByteBuffer blobBytes;
      var threeMbs = 3 * 1024 * 1024;
      if (file.exists() && file.length() > threeMbs) {
        logger.log(
            Level.FINEST,
            "Cache file {0} mmapped with {1} size",
            new Object[] {file, file.length()});
        closeMemoryArena();
        memoryArena = memoryArenaSupplier.get();
        try (var chan = FileChannel.open(file.toPath())) {
          assert memoryArena.scope().isAlive();
          var memSegment = chan.map(MapMode.READ_ONLY, 0, file.length(), memoryArena);
          assert memSegment.isReadOnly();
          blobBytes = memSegment.asByteBuffer();
        } catch (IOException e) {
          logger.log(Level.SEVERE, "Failed to mmap cache file " + file, e);
          throw e;
        }
      } else {
        blobBytes = ByteBuffer.wrap(dataPath.readAllBytes());
      }
      boolean blobDigestValid =
          !needsDataDigestVerification
              || CacheUtils.computeDigestFromBytes(blobBytes).equals(spi.blobHash(meta));

      if (sourceDigestValid && blobDigestValid) {
        try {
          long now = System.currentTimeMillis();
          var cachedObject = spi.deserialize(context, blobBytes, meta, logger);
          long took = System.currentTimeMillis() - now;
          if (cachedObject != null) {
            logger.log(
                Level.FINEST,
                "Loaded cache for {0} with {1} bytes in {2} ms",
                new Object[] {logName, blobBytes.limit(), took});
            return cachedObject;
          } else {
            invalidateCache(cacheRoot, logger);
            throw new IOException(logName + " is corrupted on disk");
          }
        } catch (ClassNotFoundException ex) {
          invalidateCache(cacheRoot, logger);
          throw new IOException("Cannot load " + dataPath + " due to: " + ex.getMessage(), ex);
        }
      } else {
        invalidateCache(cacheRoot, logger);
        throw new IOException(
            "One or more digests did not match for the cache for [" + logName + "].");
      }

    } else {
      invalidateCache(cacheRoot, logger);
      throw new FileNotFoundException(
          "Could not load the cache metadata at ["
              + toMaskedPath(metadataPath).applyMasking()
              + "].");
    }
  }

  /**
   * Close any previous arena and creates a new one. Closing the previous arena invalidates all byte
   * buffers associated with it.
   */
  private void closeMemoryArena() {
    if (memoryArena != null && memoryArena.scope().isAlive()) {
      memoryArena.close();
    }
  }

  /**
   * Read metadata representation from the provided location
   *
   * @param path location of the serialized metadata
   * @return deserialized metadata, or [[None]] if invalid
   */
  private M loadCacheMetadata(TruffleFile path, TruffleLogger logger) throws IOException {
    if (path.isReadable()) {
      return spi.metadataFromBytes(path.readAllBytes(), logger);
    } else {
      throw new FileNotFoundException("Cannot read " + path);
    }
  }

  /**
   * Gets the path to the cache data within the `cacheRoot`.
   *
   * @param cacheRoot the root of the cache for this entry
   * @return the name of the data file for this entry's cache
   */
  private TruffleFile getCacheDataPath(TruffleFile cacheRoot) {
    return cacheRoot.resolve(cacheFileName(spi.dataSuffix()));
  }

  private TruffleFile getCacheMetadataPath(TruffleFile cacheRoot) {
    return cacheRoot.resolve(cacheFileName(spi.metadataSuffix()));
  }

  /**
   * Computes the cache file name for a given extension.
   *
   * @param suffix the extension
   * @return the cache file name with the provided `ext`
   */
  private String cacheFileName(String suffix) {
    return spi.entryName() + suffix;
  }

  /**
   * Deletes the cache for this data in the provided `cacheRoot`.
   *
   * @param cacheRoot the root of the cache to delete
   * @param logger a logger
   */
  private void invalidateCache(TruffleFile cacheRoot, TruffleLogger logger) {
    TruffleFile metadataFile = getCacheMetadataPath(cacheRoot);
    TruffleFile dataFile = getCacheDataPath(cacheRoot);

    doDeleteAt(cacheRoot, metadataFile, logger);
    doDeleteAt(cacheRoot, dataFile, logger);
  }

  private void doDeleteAt(TruffleFile cacheRoot, TruffleFile file, TruffleLogger logger) {
    try {
      if (file.exists()) {
        if (file.isWritable()) {
          file.delete();
          logger.log(
              logLevel, "Invalidated the cache at [" + toMaskedPath(file).applyMasking() + "].");
        } else {
          logger.log(
              logLevel,
              "Cannot invalidate the cache at ["
                  + toMaskedPath(file).applyMasking()
                  + "]. "
                  + "Cache location not writable.");
        }
      }
    } catch (NoSuchFileException nsfe) {
      // If it doesn't exist, our work has already been done for us!
    } catch (IOException | SecurityException e) {
      logger.log(
          logLevel,
          "Unable to delete the cache at [" + toMaskedPath(cacheRoot).applyMasking() + "].");
    }
  }

  /**
   * Invalidates all caches associated with this cache.
   *
   * @param context the langage context in which loading is taking place
   */
  public final void invalidate(EnsoContext context) {
    synchronized (LOCK) {
      TruffleLogger logger = context.getLogger(this.getClass());
      for (var root : spi.getCacheRoots(context)) {
        invalidateCache(root, logger);
      }
      closeMemoryArena();
    }
  }

  final <T> T asSpi(Class<T> type) {
    return type.cast(spi);
  }

  private static boolean writeBytesTo(TruffleFile file, byte[] bytes) {
    try (OutputStream stream =
        file.newOutputStream(
            StandardOpenOption.WRITE,
            StandardOpenOption.CREATE,
            StandardOpenOption.TRUNCATE_EXISTING)) {
      stream.write(bytes);
    } catch (IOException ioe) {
      return false;
    } catch (SecurityException se) {
      return false;
    }
    return true;
  }

  private static MaskedPath toMaskedPath(TruffleFile truffleFile) {
    return new MaskedPath(Path.of(truffleFile.getPath()));
  }

  /**
   * Set of methods to be implemented by those who want to cache something.
   *
   * @param <T>
   */
  public static interface Spi<T, M> {
    /**
     * Deserializes and validates data by returning the expected cached entry, or {@code null}.
     *
     * @param context the context
     * @param data data to deserialize object from
     * @param meta metadata corresponding to the `obj`
     * @param logger Truffle's logger
     * @return {@code data} transformed to a cached entry or {@code null}
     * @throws ClassNotFoundException exception thrown on unexpected deserialized data
     * @throws IOException when I/O goes wrong
     */
    public abstract T deserialize(
        EnsoContext context, ByteBuffer data, M meta, TruffleLogger logger)
        throws IOException, ClassNotFoundException;

    /**
     * Returns the exact data to be serialized. Override in subclasses to turn an {@code entry} into
     * an array of bytes to persist
     *
     * @param context context we operate in
     * @param entry entry to persist
     * @return array of bytes
     * @throws java.io.IOException if something goes wrong
     */
    public abstract byte[] serialize(EnsoContext context, T entry) throws IOException;

    /**
     * Return serialized representation of data's metadata.
     *
     * @param sourceDigest digest of data's source
     * @param blobDigest digest of serialized data
     * @param entry data to serialize
     * @return raw bytes representing serialized metadata
     * @throws java.io.IOException in case of I/O error
     */
    public abstract byte[] metadata(String sourceDigest, String blobDigest, T entry)
        throws IOException;

    /**
     * De-serializes raw bytes to data's metadata.
     *
     * @param bytes raw bytes representing metadata
     * @param logger logger to use
     * @return non-empty metadata, if de-serialization was successful
     * @throws IOException in case of I/O error
     */
    public abstract M metadataFromBytes(byte[] bytes, TruffleLogger logger) throws IOException;

    /**
     * Compute digest of cache's data
     *
     * @param entry data for which digest should be computed
     * @param logger Truffle's logger
     * @return non-empty digest, if successful
     */
    public abstract Optional<String> computeDigest(T entry, TruffleLogger logger);

    /**
     * Compute digest of data's source
     *
     * @param context the language context in which loading is taking place
     * @param logger Truffle's logger
     * @return non-empty digest, if successful
     */
    public abstract Optional<String> computeDigestFromSource(
        EnsoContext context, TruffleLogger logger);

    /**
     * Returns locations where caches can be located
     *
     * @param context the language context in which loading is taking place
     * @return non-empty if the locations have been inferred successfully, empty otherwise
     */
    public abstract Iterable<TruffleFile> getCacheRoots(EnsoContext context);

    public abstract String entryName();

    public abstract String dataSuffix();

    public abstract String metadataSuffix();

    public abstract String sourceHash(M meta);

    public abstract String blobHash(M meta);
  }
}
